### ISSUE TITLE
Analyzer incorrectly reports "prefer_initializing_formals"

### ISSUE THREAD

**Issue by obeobe** [dart-lang/linter#3134]
Analyzer incorrectly reports "prefer_initializing_formals" in this code:

Code:
```dart
class Hello {
  bool isAdult = true;
  late int age;

  Hello({
    this.isAdult = true,
    int? age
  }) {
    age ??= this.isAdult ? 30 : 12;
    this.age = age; // <-- "prefer_initializing_formals" reported for this line
  }
}
```

Screenshot: 
![image](https://user-images.githubusercontent.com/1394039/148644071-a3703d83-7750-4758-bde1-29e1cd646c59.png)

Using:
 * Linux.
 * Dart SDK Version: `Dart SDK version: 2.15.1 (stable) (Tue Dec 14 13:32:21 2021 +0100) on "linux_x64"`
 * IntelliJ IDEA 2021.2.1
 * JetBrains Dart plugin: 212.5080.8
 * Flutter plugin: 60.1.4

**Comment by bwilkerson** [comment]
I believe that the problem here is that an initializing formal doesn't introduce a parameter into the name scope. Converting the original code to use an initializing formal therefore change the reference to `age` from being a reference to a parameter of type `int?` into a reference to a field of type `int`, which introduces errors in the body.

The most complete solution here would be to analyze the initializers and body to see whether the parameter is being referenced in such a way that making that conversion would be breaking. That might be difficult to get right. A simpler approach would be to see whether the parameter is referenced at all, and to not lint if it is, but that would likely introduce some false negatives.

**Comment by oprypin** [comment]
> see whether the parameter is referenced at all

That approach seems great to me. In fact, maybe that's even *the* important part.
If the order of occurrences is [parameter definition], [unrelated reference], [actual assignment],
the lint shouldn't change it to [parameter definition+actual assignment], [unrelated reference]
